tracer/src/Datadog.Trace/Ci/Agent/CIAgentWriter.cs

```diff
@@ -6,22 +6,55 @@
 using System;
 using System.Threading.Tasks;
 using Datadog.Trace.Agent;
-using Datadog.Trace.Ci.Tags;
+using Datadog.Trace.Ci.EventModel;
 using Datadog.Trace.Configuration;
 using Datadog.Trace.Sampling;
 
 namespace Datadog.Trace.Ci.Agent
 {
-    internal class CIAgentWriter : IAgentWriter
+    /// <summary>
+    /// CI Visibility Agent Writer
+    /// </summary>
+    internal class CIAgentWriter : IEventWriter
     {
-        private readonly AgentWriter _agentWriter = null;
-        private readonly bool _isPartialFlushEnabled = false;
+        [ThreadStatic]
+        private static Span[] _spanArray;
+        private readonly AgentWriter _agentWriter;
 
-        public CIAgentWriter(ImmutableTracerSettings settings, ISampler sampler)
+        public CIAgentWriter(ImmutableTracerSettings settings, ISampler sampler, int maxBufferSize = 1024 * 1024 * 10)
         {
-            _isPartialFlushEnabled = settings.Exporter.PartialFlushEnabled;
-            var api = new Api(settings.Exporter.AgentUri, TracesTransportStrategy.Get(settings.Exporter), null, rates => sampler.SetDefaultSampleRates(rates), _isPartialFlushEnabled);
-            _agentWriter = new AgentWriter(api, null, maxBufferSize: settings.TraceBufferSize);
+            var isPartialFlushEnabled = settings.Exporter.PartialFlushEnabled;
+            var apiRequestFactory = TracesTransportStrategy.Get(settings.Exporter);
+            var api = new Api(settings.Exporter.AgentUri, apiRequestFactory, null, rates => sampler.SetDefaultSampleRates(rates), isPartialFlushEnabled);
+            _agentWriter = new AgentWriter(api, null, maxBufferSize: maxBufferSize);
+        }
+
+        public CIAgentWriter(IApi api, int maxBufferSize = 1024 * 1024 * 10)
+        {
+            _agentWriter = new AgentWriter(api, null, maxBufferSize: maxBufferSize);
+        }
+
+        public void WriteEvent(IEvent @event)
+        {
+            // To keep compatibility with the agent version of the payload, any IEvent conversion to span
+            // goes here.
+
+            if (_spanArray is not { } spanArray)
+            {
+                spanArray = new Span[1];
+                _spanArray = spanArray;
+            }
```

Comment:
Yes, what I'm adding to that is the data locality with the local var, to avoid access multiple times to the `ThreadStatic` field. So, if the `_spanArray is not null` the value is copied to the `spanArray` local var, if null then we create the array and update the `ThreadStatic` field.

Reply:
Ooh, I see. I think we've crossed my C# pattern matching threshold 😅 And just to be clear, the array is only a field there for caching purposes, correct? It's not accessed outside of this method, so we could just new it each time. Given that, is it worth simplifying this a bit, making it `readonly` and removing the first `if`? It doesn't seem worth having the extra "new up" logic here, given it will only execute once per thread, and is virtually guaranteed to run?

Reply:
Did you mean to put the `readonly` in the `ThreadStatic` `_spanArray` ? In that case, we cannot do that, we cannot initialize a ThreadStatic field using the .ctor (only the value for the thread running the .ctor will be initialized). 

https://docs.microsoft.com/en-us/dotnet/api/system.threadstaticattribute?view=net-6.0.

> Note: Do not specify initial values for fields marked with ThreadStaticAttribute, because such initialization occurs only once, when the class constructor executes, and therefore affects only one thread. If you do not specify an initial value, you can rely on the field being initialized to its default value if it is a value type, or to null if it is a reference type.